tensorflow_encrypted/session.py

```diff
@@ -0,0 +1,135 @@
+from typing import Dict, List, Optional, Any, Union
+import os
+from collections import Iterable, defaultdict
+
+import numpy as np
+import tensorflow as tf
+from tensorflow.python.client import timeline
+
+__TFE_STATS__ = bool(os.getenv('TFE_STATS', False))
+__TFE_TRACE__ = bool(os.getenv('TFE_TRACE', False))
+__TFE_DEBUG__ = bool(os.getenv('TFE_DEBUG', False))
+__TENSORBOARD_DIR__ = str(os.getenv('TFE_STATS_DIR', '/tmp/tensorboard'))
+
+_run_counter = defaultdict(int)  # type: Any
+
+
+class TFESession():
+    """
+    Wrap a Tensorflow Session
+    """
+
+    def __init__(self, sess: tf.Session) -> None:
+        self.sess = sess
+
+    def __enter__(self) -> 'TFESession':
+        self.sess = self.sess.__enter__()
+        return self
+
+    def __exit__(self, exec_type, exec_value, exec_tb):
+        self.sess.__exit__(exec_type, exec_value, exec_tb)
+
+    def sanitize_fetches(self, fetches: Any) -> Union[List[Any], tf.Tensor, tf.Operation]:
+        if not isinstance(fetches, Iterable) or isinstance(fetches, tf.Tensor):
+            if not isinstance(fetches, tf.Tensor) and not isinstance(fetches, tf.Operation):
+                return fetches.value_on_0.backing
+            else:
+                return fetches
+
+        sanitized_fetches: List[Any] = []
+        for idx, fetch in enumerate(fetches):
+            if isinstance(fetch, Iterable) and not isinstance(fetch, tf.Tensor):
+                sanitized_fetches.append([])
+                for sub_idx, sub_fetch in enumerate(fetch):
+                    if not isinstance(sub_fetch, tf.Tensor) and not isinstance(sub_fetch, tf.Operation):
+                        sanitized_fetches[idx].append(sub_fetch.value_on_0.backing)
+                    else:
+                        sanitized_fetches[idx].append(sub_fetch)
+            else:
+                if not isinstance(fetch, tf.Tensor) and not isinstance(fetch, tf.Operation):
+                    sanitized_fetches.append(fetch.value_on_0.backing)
+                else:
+                    sanitized_fetches.append(fetch)
+
+        return sanitized_fetches
+
+    def decode_fetches(self, fetches: Any, fetches_out: Any) -> List[Union[tf.Tensor, tf.Operation]]:
```

Comment:
same here, looks like a recursive function! note remark later that you could probably decode during sanitizing

Reply:
Done